hera_cal/frf.py

```diff
@@ -0,0 +1,354 @@
+               "If rephase is True, must feed lsts and freqs"
+
+    # unwrap lsts if fed
+    if lsts is not None:
+        lsts = np.unwrap(lsts)
+
+    # form flags if None
+    if flags is None:
+        flags = np.zeros_like(data, dtype=np.bool)
+    assert isinstance(flags, np.ndarray), "flags must be fed as an ndarray"
+
+    # turn flags into weights
+    flagw = (~flags).astype(np.float)
+
+    # form nsamples if None
+    if nsamples is None:
+        nsamples = np.ones_like(data, dtype=np.float)
+    assert isinstance(nsamples, np.ndarray), "nsamples must be fed as an ndarray"
+
+    # assert Navg makes sense
+    Ntimes = data.shape[0]
+    assert Navg <= Ntimes and Navg > 0, "Navg must satisfy 0 < Navg <= Ntimes"
+
+    # calculate Navg_times, the number of remaining time samples after averaging
+    Navg_times = float(Ntimes) / Navg
+    if Navg_times % 1 > 1e-10:
+        if verbose: print "Warning: Ntimes is not evenly divisible by Navg, " \
+                          "meaning the last output time sample will be noisier "\
+                          "than the others."
+    Navg_times = int(np.ceil(Navg_times))
+
+    # form output avg list
+    avg_data = []
+    win_flags = []
+    avg_lsts = []
+    avg_nsamples = []
+    avg_extra_arrays = dict([('avg_{}'.format(a), []) for a in extra_arrays])
+
+    # iterate through Navg_times
+    for i in range(Navg_times):
+        # get starting and stopping indices
+        start = i * Navg
+        end = (i + 1) * Navg
+        d = data[start:end, :]
+        f = flags[start:end, :]
+        fw = flagw[start:end, :]
+        n = nsamples[start:end, :]
+
+        # calculate mean_l and l, if lsts was fed
+        if lsts is not None:
+            l = lsts[start:end]
+            mean_l = np.mean(l)
+            avg_lsts.append(mean_l)
+
+        # rephase data if desired
+        if rephase:
+            # get dlst and rephase
+            dlst = mean_l - l
+            d = utils.lst_rephase(d, bl_vec, freqs, dlst, lat=lat, inplace=False, array=True)
+
+        w = fw * n
+        w_sum = np.sum(w, axis=0, keepdims=False).clip(1e-10, np.inf)
+
+        ad = np.sum(d * w, keepdims=False, axis=0) / w_sum
+        an = np.sum(w, keepdims=False, axis=0)
+
+        avg_data.append(ad)
+        win_flags.append(np.min(f, axis=0, keepdims=False))
+        avg_nsamples.append(an)
+
+        # average arrays in extra_arrays
+        for a in extra_arrays:
+            avg_extra_arrays['avg_{}'.format(a)].append(np.mean(extra_arrays[a][start:end]))
+
+    avg_data = np.array(avg_data, np.complex)
+    win_flags = np.array(win_flags, np.bool)
+    avg_nsamples = np.array(avg_nsamples, np.float)
+    avg_lsts = np.array(avg_lsts, np.float)
+
+    # wrap lsts
+    avg_lsts = avg_lsts % (2*np.pi)
+
+    return avg_data, win_flags, avg_nsamples, avg_lsts, avg_extra_arrays
+
+
+class FRFilter(object):
+    """
+    Fringe Rate Filter
+    """
+    def __init__(self):
```

Comment:
there currently isn't anything done in the `__init__`, but i wanted to follow the same structure as `Delay_Filter`, especially if this module grows to incorporate a true fringe rate filter and become more complicated, it would be nice to have this kind of structure